### PR TITLE
Apply immunity-cache to SCR Pool and Rewards Pool

### PR DESCRIPTION
**Pull request by andreibancioiu**
1. Use `ImmunityCache` instead of `FIFOShardedCache` for `shardedData` (SCR Pool & Rewards Pool)
1. Pass `SizeInBytes` cleanly (as received in `AddData()`) to `TxCache` instead of estimating it
1. Update some tests where `TxPool` was still using LRU. `ShardedTxPool` was actually working nonetheless :)
1. Update `Cacher.HasOrAdd()` to return `(has, added bool)` - the output parameter `evicted` wasn't used - and it should be opaque to callers anyway.
1. In `ImmunityCache`, remove the interface `CacheItem`. Accept and wrap any kind of items added to the cache.
1. Move common mocks and stubs for cache and pool to a new package, `testscommon` - and thus removed some duplication
1. Remove some unused code

Gathered some remaining (old or new) TODOs here: [EN-6732](https://elrondnetwork.atlassian.net/browse/EN-6732)

### CODE REVIEW

File: dataRetriever/shardedData/shardedData.go [hunk 1]
```diff
@@ -16,109 +18,86 @@ var _ dataRetriever.ShardedDataCacherNotifier = (*shardedData)(nil)
...
 // NewShardedData is responsible for creating an empty pool of data
-func NewShardedData(cacherConfig storageUnit.CacheConfig) (*shardedData, error) {
-	err := verifyCacherConfig(cacherConfig)
-	if err != nil {
-		return nil, err
+func NewShardedData(name string, config storageUnit.CacheConfig) (*shardedData, error) {
+	log.Info("NewShardedData", "name", name, "config", config.String())
+
+	configPrototype := immunitycache.CacheConfig{
+		Name:                        "prototype",
```
Comment (sasurobert) [inline dataRetriever/shardedData/shardedData.go:48]: magic name

File: storage/immunitycache/chunk.go [hunk 2]
```diff
@@ -129,9 +129,9 @@ func (chunk *immunityChunk) removeOldestNoLock(numToRemove int) int {
 	element := chunk.itemsAsList.Front()
 
 	for element != nil && numRemoved < numToRemove {
-		item := element.Value.(storage.CacheItem)
+		item := element.Value.(*cacheItem)
```
Comment (sasurobert) [inline storage/immunitycache/chunk.go:132]: check with ok the type assertion return if wrong

File: storage/immunitycache/chunk.go [hunk 3]
```diff
@@ -147,8 +147,8 @@ func (chunk *immunityChunk) removeOldestNoLock(numToRemove int) int {
 }
 
 func (chunk *immunityChunk) removeNoLock(element *list.Element) {
-	item := element.Value.(storage.CacheItem)
-	delete(chunk.items, string(item.GetKey()))
+	item := element.Value.(*cacheItem)
+	delete(chunk.items, item.key)
```
Comment (sasurobert) [inline storage/immunitycache/chunk.go:151]: check with ok the type assertion return if wrong. same for every type assertion.

File: dataRetriever/shardedData/shardedData.go [hunk 4]
```diff
@@ -127,71 +109,103 @@ func (sd *shardedData) AddData(key []byte, value interface{}, sizeInBytes int, c
...
+	log.Debug("shardedData.removeTxBulk()", "name", sd.name, "cacheID", cacheID, "numToRemove", len(keys), "numRemoved", numRemoved)
 }
 
-// ImmunizeSetOfDataAgainstEviction does nothing
-func (sd *shardedData) ImmunizeSetOfDataAgainstEviction(_ [][]byte, _ string) {
+// ImmunizeSetOfDataAgainstEviction  marks the items as non-evictable
+func (sd *shardedData) ImmunizeSetOfDataAgainstEviction(keys [][]byte, cacheID string) {
+	store := sd.shardStore(cacheID)
+	if store == nil {
+		return
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply dataRetriever/shardedData/shardedData.go:174]: Fixed.

**Pull request merged**